### PR TITLE
Unit tests for Map Options

### PR DESCRIPTION
**Pull request by jakublala**
Made a few unit tests for Map Options - it's primarily checking selecting/clicking on the elements. Please, have a look if the unit tests make sense, or if I should do them differently.

Happy to do some more, but I would appreciate some guidance - let's say a list of things we want and can test with this basic framework setup.

### CODE REVIEW

**Comment by Luthaf** [comment]
Could you rebase this branch now that the karma framework is merged?

**Comment by jakublala** [comment]
Rebased and added/finished a color reset unit test. I am happy to leave this PR open, but I am not sure whether I will be eager to add more unit tests (in the near future), so it might be good just to get it approved and merge it - so that it lives in the repository as a good 'guidance'/template for other people collaborating and adding their own unit tests.

File: tests/map/map.test.ts [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+        const root = document.createElement('div');
+        const indexer = new EnvironmentIndexer('structure', DUMMY_STRUCTURES);
+        const map = new PropertiesMap(
+            { element: root, settings: DUMMY_MAP_SETTINGS },
+            indexer,
+            DUMMY_PROPERTIES
+        );
+
+        assert(root.innerHTML !== '');
+        assert(document.body.innerHTML !== '');
```
Comment (Luthaf) [inline tests/map/map.test.ts:85]:
```suggestion
        assert(document.body.innerHTML !== KARMA_INSERTED_HTML);
```

File: tests/map/map.test.ts [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+
+        map.remove();
+        assert(root.innerHTML === '');
+
+        // remove SVG element created by Plotly
```
Comment (Luthaf) [inline tests/map/map.test.ts:90]: this should be part of the map `remove` function

File: tests/map/map.test.ts [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+        resetButton.dispatchEvent(new window.Event('click'));
+
+        assert(options.color.min.value === originalMin);
+        assert(options.color.max.value === originalMax);
+
+        removeMap(map);
+    });
+});
+
+/** Removes the map and an additional HTML element creted by Plotly */
```
Comment (Luthaf) [inline tests/map/map.test.ts:129]: this function does not seems to be used

File: tests/map/options.test.ts [hunk 4]
```diff
@@ -33,7 +36,7 @@ describe('MapOptions', () => {
         assert(document.body.innerHTML === KARMA_INSERTED_HTML);
     });
 
-    it('scale label for min/max switches between linear and log', () => {
+    it('selected scale label for min/max switches between linear and log', () => {
```
Comment (Luthaf) [inline tests/map/options.test.ts:39]: Not sure why this change is here, the previous test name was clearer IMO

File: tests/map/options.test.ts [hunk 5]
```diff
@@ -61,4 +64,63 @@ describe('MapOptions', () => {
...
+        const options = new MapOptions(root, DUMMY_PROPERTIES, DUMMY_CALLBACK);
+
+        checkPropertySelect(options.x, 'x');
+        checkPropertySelect(options.y, 'y');
+        checkPropertySelect(options.z, 'z');
+        checkPropertySelect(options.color, 'color');
+
+        function checkPropertySelect(axisOptions: AxisOptions, axisName: string) {
+            const selectElement = getByID<HTMLSelectElement>(`chsp-${axisName}`);
+            selectElement.value = 'second';
```
Comment (Luthaf) [inline tests/map/options.test.ts:79]: We want to check that the value changes, so we need to check that the new & previous value are different, with something like this:

```suggestion
            assert(selectElement.value !== 'second');
            assert(axisOptions.property.value !== 'second');
            selectElement.value = 'second';
```

File: tests/map/options.test.ts [hunk 6]
```diff
@@ -61,4 +64,63 @@ describe('MapOptions', () => {
...
+            selectElement.dispatchEvent(new window.Event('change'));
+            assert(axisOptions.property.value === 'second');
+        }
+    });
+
+    it('selected range changes when changed', () => {
```
Comment (Luthaf) [inline tests/map/options.test.ts:85]:
```suggestion
    it('axis range changes when min/max are modified', () => {
```

File: tests/map/options.test.ts [hunk 7]
```diff
@@ -61,4 +64,63 @@ describe('MapOptions', () => {
             assert(max.innerHTML === 'max:');
         }
     });
+
+    it('selected property changes when changed', () => {
```
Comment (Luthaf) [inline tests/map/options.test.ts:68]:
```suggestion
    it('axis property changes when modified by the user', () => {
```

File: tests/map/options.test.ts [hunk 8]
```diff
@@ -61,4 +64,63 @@ describe('MapOptions', () => {
...
+        }
+    });
+
+    it('selected range changes when changed', () => {
+        const root = document.createElement('div');
+        const options = new MapOptions(root, DUMMY_PROPERTIES, DUMMY_CALLBACK);
+
+        checkRangeSelect(options.x, 'x');
+        checkRangeSelect(options.y, 'y');
+        checkRangeSelect(options.z, 'z');
```
Comment (Luthaf) [inline tests/map/options.test.ts:91]: why not the color here?

File: tests/map/options.test.ts [hunk 9]
```diff
@@ -61,4 +64,63 @@ describe('MapOptions', () => {
...
+
+        function checkRangeSelect(axisOptions: AxisOptions, axisName: string) {
+            const minSelectElement = getByID<HTMLSelectElement>(`chsp-${axisName}-min`);
+            const maxSelectElement = getByID<HTMLSelectElement>(`chsp-${axisName}-max`);
+
+            minSelectElement.value = '0.01';
```
Comment (Luthaf) [inline tests/map/options.test.ts:97]: same as above, this needs to check the previous value as well

File: tests/map/options.test.ts [hunk 10]
```diff
@@ -61,4 +64,63 @@ describe('MapOptions', () => {
...
+            assert(axisOptions.max.value === 1.01);
+        }
+    });
+
+    it('selected size changes when changed', () => {
+        const root = document.createElement('div');
+        const options = new MapOptions(root, DUMMY_PROPERTIES, DUMMY_CALLBACK);
+        const selectElement = getByID<HTMLSelectElement>(`chsp-size`);
+
+        selectElement.value = 'second';
```
Comment (Luthaf) [inline tests/map/options.test.ts:112]: same as above, this needs to check the previous value as well

File: tests/map/options.test.ts [hunk 11]
```diff
@@ -61,4 +64,63 @@ describe('MapOptions', () => {
...
+        selectElement.dispatchEvent(new window.Event('change'));
+        assert(options.size.property.value === 'second');
+    });
+
+    it('slider size changes when changed', () => {
+        const root = document.createElement('div');
+        const options = new MapOptions(root, DUMMY_PROPERTIES, DUMMY_CALLBACK);
+        const sliderElement = getByID<HTMLInputElement>(`chsp-size-factor`);
+
+        sliderElement.value = '100';
```
Comment (Luthaf) [inline tests/map/options.test.ts:122]: same as above, this needs to check the previous value as well

File: tests/map/options.test.ts [hunk 12]
```diff
@@ -61,4 +64,63 @@ describe('MapOptions', () => {
...
+            minSelectElement.dispatchEvent(new window.Event('change'));
+            assert(axisOptions.min.value === 0.01);
+
+            maxSelectElement.value = '1.01';
+            maxSelectElement.dispatchEvent(new window.Event('change'));
+            assert(axisOptions.max.value === 1.01);
+        }
+    });
+
+    it('selected size changes when changed', () => {
```
Comment (Luthaf) [inline tests/map/options.test.ts:107]:
```suggestion
    it('property used for size changes when modified by the user', () => {
```

File: tests/map/options.test.ts [hunk 13]
```diff
@@ -61,4 +64,63 @@ describe('MapOptions', () => {
...
+        const root = document.createElement('div');
+        const options = new MapOptions(root, DUMMY_PROPERTIES, DUMMY_CALLBACK);
+        const selectElement = getByID<HTMLSelectElement>(`chsp-size`);
+
+        selectElement.value = 'second';
+        selectElement.dispatchEvent(new window.Event('change'));
+        assert(options.size.property.value === 'second');
+    });
+
+    it('slider size changes when changed', () => {
```
Comment (Luthaf) [inline tests/map/options.test.ts:117]:
```suggestion
    it('size factor changes when the slider is modified by the user', () => {
```

File: tests/map/options.test.ts (hunk 4 above)
Comment (jakublala) [reply tests/map/options.test.ts:39]: i have changed it to 'selected' so that it aligns with the other tests, where a 'selection' is always made on the HTML select element

Comment (jakublala) [reply tests/map/options.test.ts:39]: but seeing that you changed the others to something clearer, then i will revert this one

File: tests/map/options.test.ts (hunk 8 above)
Comment (Luthaf) [reply tests/map/options.test.ts:95]: Still an open question =)

File: tests/map/map.test.ts [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+        minSelectElement.dispatchEvent(new window.Event('change'));
+        assert(options.color.min.value === 1.0);
+
+        maxSelectElement.value = '3.0';
+        maxSelectElement.dispatchEvent(new window.Event('change'));
+        assert(options.color.max.value === 3.0);
+
+        const resetButton = map['_colorReset'];
+        resetButton.click();
+        resetButton.dispatchEvent(new window.Event('click'));
```
Comment (Luthaf) [inline tests/map/map.test.ts:118]: Why do we need both? `click()` does not propagate the event?

Comment (jakublala) [reply tests/map/map.test.ts:118]: nope. click() itself does not propagate and one needs to dispatch the event for the callback function to be called

File: tests/map/options.test.ts (hunk 8 above)
Comment (jakublala) [reply tests/map/options.test.ts:95]: sorry, i didn't notice it earlier. i just added a new commit

**Review by Luthaf — COMMENTED** [review]
Looks good, given that I just merged #132 I'll manually rebase this PR to check it locally before merging

**Review by Luthaf — APPROVED** [review]
